### ISSUE TITLE
Drop dependency cordova-plugin-file-transfer

### ISSUE THREAD

**Issue by kaynz** [Microsoft/cordova-plugin-code-push#341]
Hi!

### Description
The plugin cordova-plugin-file-transfer has recently been deprecated. Additionally it's causing errors if in the same project cordova-plugin-file ^6.0.0 is installed.

### Reproduction

Install cordova-plugin-file@latest
Install cordova-plugin-file-transfer@latest

### Additional Information

* cordova-plugin-code-push version: latest
* List of installed plugins: cordova-plugin-file:6.0.1
* Cordova version: 8.0.0
* iOS/Android version: 4.5.4/7.0.0

### Possible solutions

* Drop usage of cordova-plugin-file-transfer. Follow the migration guide and upgrade to XHR: https://cordova.apache.org/blog/2017/10/18/from-filetransfer-to-xhr2.html
* Fork cordova-plugin-file-transfer and raise version to ^6.0.0 (pretty good chance that new bugs will appear)
* Believe in god and hope that cordova merges https://github.com/apache/cordova-plugin-file-transfer/pull/195. Update to that new version in cordova-plugin-code-push.

**Comment by sergey-akhalkov** [comment]
Hi @kaynz, thanks for reaching us! We've recently released [1.11.0](https://github.com/Microsoft/cordova-plugin-code-push/releases/tag/v1.11.0) version of Cordova Plugin for CodePush, so could you please update the plugin version in your dependencies and let us know if it resolves the issue?

**Comment by fikriauliya** [comment]
I encounter this issue in 1.11.0

```
Discovered plugin "cordova-plugin-code-push" in config.xml. Adding it to the project
Installing "cordova-plugin-code-push" for ios
Installing "code-push" for ios
Installing "cordova-plugin-file" for ios
Installing "cordova-plugin-file-transfer" for ios
Failed to install 'cordova-plugin-file-transfer': CordovaError: Version of installed plugin: "cordova-plugin-file@6.0.1" does not satisfy dependency plugin requirement "cordova-plug
in-file@^5.0.0". Try --force to use installed plugin as dependency.
    at /Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:557:37
    at _fulfilled (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:749:13)
    at /Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:509:49
    at flush (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:108:17)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
Failed to install 'cordova-plugin-code-push': CordovaError: Version of installed plugin: "cordova-plugin-file@6.0.1" does not satisfy dependency plugin requirement "cordova-plugin-f
ile@^5.0.0". Try --force to use installed plugin as dependency.
    at /Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:557:37
    at _fulfilled (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:749:13)
    at /Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:509:49
    at flush (/Users/levi/.nvm/versions/node/v8.9.3/lib/node_modules/cordova/node_modules/cordova-lib/node_modules/q/q.js:108:17)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
Failed to restore plugin "cordova-plugin-code-push" from config.xml. You might need to try adding it again. Error: Version of installed plugin: "cordova-plugin-file@6.0.1" does not
satisfy dependency plugin requirement "cordova-plugin-file@^5.0.0". Try --force to use installed plugin as dependency.
```

**Comment by kaynz** [comment]
@sergey-akhalkov That's what I did. :)

**Comment by DLopezOpentix** [comment]
Same issue here:

`cordova plugin add platform ios`

```
Installing "cordova-plugin-code-push" for ios

Plugin dependency "code-push@2.0.5" already fetched, using that version.

Dependent plugin "code-push" already installed on ios.

Installing "cordova-plugin-file-transfer" for ios
Failed to install 'cordova-plugin-file-transfer': CordovaError: Version of installed plugin: "cordova-plugin-file@6.0.1" does not satisfy dependency plugin requirement "cordova-plugin-file@^4.0.0". Try --force to use installed plugin as dependency.
    at /usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:581:33
    at _fulfilled (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:749:13)
    at /usr/local/lib/node_modules/cordova/node_modules/q/q.js:509:49
    at flush (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:108:17)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)

Failed to install 'cordova-plugin-code-push': CordovaError: Version of installed plugin: "cordova-plugin-file@6.0.1" does not satisfy dependency plugin requirement "cordova-plugin-file@^4.0.0". Try --force to use installed plugin as dependency.
    at /usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:581:33
    at _fulfilled (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:749:13)
    at /usr/local/lib/node_modules/cordova/node_modules/q/q.js:509:49
    at flush (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:108:17)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
Error: Version of installed plugin: "cordova-plugin-file@6.0.1" does not satisfy dependency plugin requirement "cordova-plugin-file@^4.0.0". Try --force to use installed plugin as dependency.
[ERROR] Exception:
```

**Comment by annakocheshkova** [comment]
@kaynz @fikriauliya @DLopezOpentix 
The thing is we have to support backwards compatibility with lower cordova versions. This is why in our plugin we require `cordova-plugin-file-transfer` to be of version 1.6.3 and therefore `cordova-plugin-file` to be of version 5.0.0.
We see that our workaround doesn't work for those who already have plugin with higher version installed. 

Right now I am working on the solution that is going to be more flexible and allow code push plugin to work out-of-the-box. This is a post-install script that determines cordova version and installs the corresponding versions of `file` and `file-transfer` plugins, thus allowing to get rid of `file-transfer` dependency in `plugin.xml`. I have created a [https://github.com/Microsoft/cordova-plugin-code-push/pull/345](PR) with these changes. Please try to see if it works.

**As a simpler temporary solution, you can do the following:** 
`cordova plugin remove cordova-plugin-file`
`cordova plugin add cordova-plugin-code-push`
`cordova plugin add cordova-plugin-file@latest`.

Also, in the future we are planning to get rid of file-transfer plugin completely as it got deprecated.

**Comment by Dellos7** [comment]
Hello @annakocheshkova ,

I've tried your workaround and it worked for me.

Thank you!

**Comment by sergey-akhalkov** [comment]
Hi @Dellos7, nice to hear! We are going to make a release at the end of this or in the beginning of the next week.

**Comment by arcanys-ivan** [comment]
no updates so far? :( i came across this issue as well as i am converting ionic1 to ionic3 which is in higher cordova versions

**Comment by kaynz** [comment]
@arcanys-ivan As you can see in https://github.com/Microsoft/cordova-plugin-code-push/pull/345 there has been a huge update recently.

**Comment by max-mironov** [comment]
Hey folks, we've released new version https://github.com/Microsoft/cordova-plugin-code-push/releases/tag/v1.11.1 for plugin, closing the issue so far, please let us know if any questions.

**Comment by bytenik** [comment]
Perhaps I'm missing something, but the dependency is still there and I still cannot build.

**Comment by bytenik** [comment]
As per https://github.com/Microsoft/cordova-plugin-code-push/blob/master/plugin.xml#L17 the dependency is still there. I thought this was moving to a postinstall script. Was there an issue with that or is there some way I can help move this along? My build is broken due to this issue.

**Comment by iTOYS** [comment]
Hi @bytenik,
`postinstall` script didn't work correctly, so we've returned dependencies in plugin.xml
I'm working on workaround

**Comment by jpike88** [comment]
@iTOYS so why exactly isn't this issue reopened?

**Comment by Zvika88** [comment]
hello @annakocheshkova,
The solution you proposed does not seem to work anymore. When trying to add cordova-plugin-code-push (in step #2), it tries to bring in "cordova-plugin-compat" for android, which then fails.

I am using android SDK API version 27. 
The use of cordova-plugin-compat and forcing cordrova-plugin-file worked if the SDK uses API 24, but that is too old and need to be changed.

How can we install code-push and NOT bring it cordova-plugin-compat?

Thanks,
Z.

**Comment by jpike88** [comment]
They haven't even reopened this issue, but @iTOYS took the time to comment that the issue is still present. I really don't get it.

**Comment by alexandergoncharov** [comment]
Hi everybody,

I'm working on it.

**Comment by iTOYS** [comment]
Will be release soon

**Comment by betocantu93** [comment]
Hello, I need to use cordova-plugin-file version="^6.0.0" and cordova-android 7.0.0. I'm not quite sure how the afterPluginAdd.js has to be modified for this to work. 

Reproduction steps:
1. `$cordova plugin add cordova-plugin-code-push@latest`
2. Manually delete cordova-plugin-file directoy
3. Manually delete cordova-plugin-file-transfer directory
4. Modify cordova-plugin-file version in afterPluginAdd.js
From 
`var output = execSync(cordovaCLI + ' plugin add cordova-plugin-file@4.3.3').toString();`
To
`var output = execSync(cordovaCLI + ' plugin add cordova-plugin-file@latest').toString();`

Output

```
Adding the cordova-plugin-file-transfer@1.6.3... 
(node:12843) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
npm http GET https://registry.npmjs.org/cordova-plugin-file-transfer/1.6.3
npm http 200 https://registry.npmjs.org/cordova-plugin-file-transfer/1.6.3
npm WARN deprecated cordova-plugin-file-transfer@1.6.3: This plugin has been deprecated
npm http GET https://registry.npmjs.org/cordova-plugin-file-transfer/-/cordova-plugin-file-transfer-1.6.3.tgz
npm http 200 https://registry.npmjs.org/cordova-plugin-file-transfer/-/cordova-plugin-file-transfer-1.6.3.tgz
Failed to install 'cordova-plugin-file-transfer':Error: ENOENT: no such file or directory, open '/home/betocantu93/Documents/apps/app_test/corber/cordova/platforms/android/AndroidManifest.xml'
    at Error (native)
    at Object.fs.openSync (fs.js:642:18)
    at Object.fs.readFileSync (fs.js:510:33)
    at Object.parseElementtreeSync (/usr/share/cordova-cli/node_modules/cordova-lib/src/util/xml-helpers.js:118:27)
    at Object.package_name (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/platforms/android.js:62:32)
    at PlatformMunger.generate_plugin_config_munge (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:264:54)
    at PlatformMunger.add_plugin_changes (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:184:29)
    at /usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:315:14
    at Array.forEach (native)
    at PlatformMunger_process [as process] (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:314:45)
Error: ENOENT: no such file or directory, open '/home/betocantu93/Documents/apps/app_test/corber/cordova/platforms/android/AndroidManifest.xml'
    at Error (native)
    at Object.fs.openSync (fs.js:642:18)
    at Object.fs.readFileSync (fs.js:510:33)
    at Object.parseElementtreeSync (/usr/share/cordova-cli/node_modules/cordova-lib/src/util/xml-helpers.js:118:27)
    at Object.package_name (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/platforms/android.js:62:32)
    at PlatformMunger.generate_plugin_config_munge (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:264:54)
    at PlatformMunger.add_plugin_changes (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:184:29)
    at /usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:315:14
    at Array.forEach (native)
    at PlatformMunger_process [as process] (/usr/share/cordova-cli/node_modules/cordova-lib/src/plugman/util/config-changes.js:314:45)
```
I noticed cordova-plugin-file-transfer@1.6.3 have

`<dependency id="cordova-plugin-file" version="^4.0.0" />`

So I repeated the above steps with also updating the cordova-plugin-file-transfer version to 1.7.1 in afterPluginAdd.js which has:

`<dependency id="cordova-plugin-file" version=">=5.0.0" />`

And I still got the same output.

So basically I just need a little bit guidance in how to keep cordova-plugin-file@latest. With of course the cordova-plugin-file-transfer not breaking my build

**Comment by dv-alan** [comment]
Any word on this? Keeps breaking our continuous integration cycle.

**Comment by yuri-kulikov** [comment]
Hi @dv-alan, @betocantu93,
Sorry for the delay!

Maybe you can provide us some simple reproducible demo so we can understand your environment better?
And I think it would be better to create a new issue to get detailed information and solve the problem.

**Comment by omardoma** [comment]
Hey, I am facing this issue as well, `cordova-plugin-file-transfer@1.6.3` is making my life a nightmare, any workarounds for this until the new version is released ?

**Comment by jackie-d** [comment]
Hi, still trouble with this nowadays.

In my case, due to cordova-plugin-media dependencies incompatibility (amenable to cordova-plugin-file).

```
cordova plugin remove cordova-plugin-media
... OK

cordova plugin add cordova-plugin-code-push
... OK

cordova plugin add cordova-plugin-media      
Installing "cordova-plugin-media" for android
Failed to install 'cordova-plugin-media': CordovaError: Version of installed plugin: "cordova-plugin-file@4.3.3" does not satisfy dependency plugin requirement "cordova-plugin-file@^6.0.0". Try --force to use installed plugin as dependency.
    at /usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:544:43
    at processTicksAndRejections (internal/process/task_queues.js:86:5)
Version of installed plugin: "cordova-plugin-file@4.3.3" does not satisfy dependency plugin requirement "cordova-plugin-file@^6.0.0". Try --force to use installed plugin as dependency.

cordova plugin add cordova-plugin-file@latest
Plugin "cordova-plugin-file" already installed on android.
Plugin "cordova-plugin-file" already installed on browser.
Plugin "cordova-plugin-file" already installed on ios.
Adding cordova-plugin-file to package.json

```

Any workaround to have `cordova-plugin-code-push` standalone or at least relying on latest `cordova-plugin-file@^6.0.0`?

**Comment by jackie-d** [comment]
Related:

https://github.com/microsoft/cordova-plugin-code-push/pull/513
https://github.com/microsoft/cordova-plugin-code-push/issues/485

**Comment by jackie-d** [comment]
Found a solution (for my specific case but works generically):

```
# clean previous state
# cordova plugin remove cordova-plugin-file #not needed in my case
cordova plugin remove cordova-plugin-media
cordova plugin remove cordova-plugin-code-push
# install needed plugins
cordova plugin add cordova-plugin-media
cordova plugin add cordova-plugin-code-push
# fix broken dependency for code-push
cordova plugin add cordova-plugin-file-transfer --force
```

**Comment by azens1995** [comment]
Even after more than 2 years, I still faced the same issue. `cordova-plugin-file-transfer` was depending on the `cordova-plugin-file` and the one that was installed on my project was greater that 6.0.0 version.

This caused the failure of the successful add of code-push plugin.

@jackie-d mentioned the solution and I even tried to implement it, but with no success.

So here's what I did. I don't know if it will be able to help anyone, but if you have similar scenario like mine, then hope this solution will help.

I tried to remove the `cordova-plugin-file` but could not as other installed plugin was depending on it. The other plugin was `cordova-plugin-advanced-http`.

So, I uninstalled the `cordova-plugin-advanced-http` first, and later the `cordova-plugin-file`. 
After that, I added the code-push plugin.

Code-push plugin automatically added the `cordova-plugin-file` with it's respective dependent version i.e. `^4.0.0`.

After that I again added the `cordova-advanced-http` plugin, which later used the same earlier installed `cordova-plugin-file` 's version.
